### PR TITLE
[Console] Add interface for command class

### PR DESCRIPTION
**Pull request by HeathNaylor**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #27665 
| License       | MIT
| Doc PR        | No documentation change

Not sure this constitutes as a feature. This is simply adding an interface for the Command class. I was also a little confused, the contribution documentation says to target version 2.7 and on, so this PR is for 2.7.

### CODE REVIEW

**Comment by HeathNaylor** [comment]
The checks are failing for fabbot.io on common typos that are outside the scope of my PR.

File: CHANGELOG-2.7.md [hunk 1]
```diff
@@ -7,6 +7,8 @@ in 2.7 minor versions.
 To get the diff for a specific change, go to https://github.com/symfony/symfony/commit/XXX where XXX is the change hash
 To get the diff between two versions, go to https://github.com/symfony/symfony/compare/v2.7.0...v2.7.1
 
+ * feature #27665 [Console] Add contract for Command class
```
Comment (iltar) [inline CHANGELOG-2.7.md]: This file does not need to be updated, it's autogenerated on release I believe

File: src/Symfony/Component/Console/Command/CommandInterface.php [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+use Symfony\Component\Console\Application;
+use Symfony\Component\Console\Helper\HelperSet;
+use Symfony\Component\Console\Input\InputDefinition;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+/**
+ * Base class for all commands.
+ */
+interface CommandInterface
```
Comment (iltar) [inline src/Symfony/Component/Console/Command/CommandInterface.php]: Doc above here is incorrect now. 

I also think this interface has far too many methods that are not relevant to implementing your command, they are merely a detail within Symfony. I would love to see this interface only define the methods you'd need to implement for a simple executable command. Perhaps multiple interfaces if you wish to implement additional features? But several of these methods are very specific to the architecture of the Command as implemented by Symfony.

File: src/Symfony/Component/Console/Command/CommandInterface.php [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+     * The code to execute is either defined directly with the
+     * setCode() method or by overriding the execute() method
+     * in a sub-class.
+     *
+     * @return int The command exit code
+     *
+     * @throws \Exception When binding input fails. Bypass this by calling {@link ignoreValidationErrors()}.
+     *
+     * @see setCode()
+     * @see execute()
```
Comment (iltar) [inline src/Symfony/Component/Console/Command/CommandInterface.php]: The references to execute are no longer  valid, as this is a detail of the abstract command

File: src/Symfony/Component/Console/Command/CommandInterface.php [hunk 4]
```diff
@@ -0,0 +1,305 @@
...
+     * @throws \InvalidArgumentException if the helper is not defined
+     */
+    public function getHelper($name);
+
+    /**
+     * Returns a text representation of the command.
+     *
+     * @return string A string representing the command
+     *
+     * @deprecated since version 2.3, to be removed in 3.0.
```
Comment (iltar) [inline src/Symfony/Component/Console/Command/CommandInterface.php]: It see you've based your PR on 2.7. 2.7 is not supported anymore and new features go into the master. Can you change your PR to be against the master instead? Github will let you change it, but you will still need to change the base branch in your git branch.

**Review by linaori — CHANGES_REQUESTED** [review]

**Comment by stof** [comment]
New features are never going into a maintenance branch. This cannot go anywhere else than master.

**Comment by nicolas-grekas** [comment]
> contribution documentation says to target version 2.7 and on, so this PR is for 2.7.

can you link to the page telling so please? This is a mistake we need to fix.

About the proposed change, I'm :-1: as it doesn't make sense to add this interface which does so many things and breaks e.g. SRP, etc. Just use the class instead.

**Comment by chalasr** [comment]
I’m -1 for the proposed change, you probably don’t want to implement this interface :)
Closing, thank you for the PR.

**Comment by HeathNaylor** [comment]
@nicolas-grekas https://symfony.com/doc/current/contributing/code/patches.html References 2.7 as the branch to use.

**Pull request closed without merging**